mail-server/borgbackup.nix

```diff
@@ -0,0 +1,72 @@
+#  nixos-mailserver: a simple mail server
+#  Copyright (C) 2016-2018  Robin Raymond
+#
+#  This program is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program. If not, see <http://www.gnu.org/licenses/>
+
+{ config, pkgs, lib, ... }:
+
+let
+  cfg = config.mailserver.borgbackup;
+  cronTimes = cfg.cronTimes;
+
+  methodFragment = lib.optional (cfg.compression.method != null) cfg.compression.method;  
+  autoFragment =
+    if cfg.compression.auto && cfg.compression.method == null
+    then throw "compression.method must be set when using auto."
+    else lib.optional cfg.compression.auto "auto";
+  levelFragment =
+    if cfg.compression.level != null && cfg.compression.method == null
+    then throw "compression.method must be set when using compression.level."
+    else lib.optional (cfg.compression.level != null) (toString cfg.compression.level);
+  compressionFragment = lib.concatStringsSep "," (lib.flatten [autoFragment methodFragment levelFragment]);
+  compression = lib.optionalString (compressionFragment != "") "--compression ${compressionFragment}";
+
+  encryptionFragment = cfg.encryption.method;
+  passphraseFile = lib.escapeShellArg cfg.encryption.passphraseFile;
+  passphraseFragment = lib.optionalString (cfg.encryption.method != "none")
+                         (if cfg.encryption.passphraseFile != null then ''env BORG_PASSPHRASE="$(cat ${passphraseFile})"''
+                          else throw "passphraseFile must be set when using encryption.");
+
+  locations = lib.escapeShellArgs cfg.locations;
+  name = lib.escapeShellArg cfg.name;
+
+  repoLocation = lib.escapeShellArg cfg.repoLocation;
+
+  extraInitArgs = lib.escapeShellArgs cfg.extraArgumentsForInit;
+  extraCreateArgs = lib.escapeShellArgs cfg.extraArgumentsForCreate;
+
+  cmdPreexec = lib.optionalString (cfg.cmdPreexec != null) cfg.cmdPreexec;
+  cmdPostexec = lib.optionalString (cfg.cmdPostexec != null) cfg.cmdPostexec;
+
+  borgScript = ''
+    . /etc/profile
```

Comment:
This should not be necessary.

Reply:
Without it it says that ionice is an unknown command:
`-start: line 5: ionice: command not found`

This is recommended by the NixOS wiki for cron jobs: https://nixos.wiki/wiki/Cron

Reply:
Alright. But if you use `serviceConfig.Nice`, you can definitely leave it out.

Reply:
How does the systemd directive `Nice` affect I/O?

Reply:
It doesn't. Sorry.
We should use `IOSchedulingClass` and `IOSchedulingPriority`: https://www.freedesktop.org/software/systemd/man/systemd.exec.html

Reply:
I went ahead and changed both `nice` and `ionice` into `CPUSchedulingPolicy` and `IOSchedulingClass` set to `idle` which is 

> for running very low priority background jobs

according to the man page of [`sched_setscheduler`](http://man7.org/linux/man-pages/man2/sched_setscheduler.2.html) and because

> [p]rocesses running at this level get I/O time only when no one else needs the disk

according to the man page of [`ioprio_set`](http://man7.org/linux/man-pages/man2/ioprio_set.2.html).

I am a bit at a loss as to what that means exactly in terms of nice value and process priority but finding out about the inner workings is a bit of a [science ](http://www.learnlinux.org.za/courses/build/internals/ch07s02.html) on its own and ultimately depends on the scheduler used. However I seem to not be [alone](https://unix.stackexchange.com/questions/340283/using-and-understanding-systemd-scheduling-related-options-in-a-desktop-context).

In my observation it sets a process to a nice level of `2` which results in a process priority of `22` most of the time. I have used the rather unscientific method of just watching how htop colours the cpu usage meters and found that a `nice` level of `2` makes the bar just as blue as a higher `nice` level. And blue means low priority. Ultimately I rely on the `man` page and if it says this is the thing to do for low priority background jobs then that's how it's gonna be.